pkg/profiler/pool.go

```diff
@@ -30,89 +30,84 @@ import (
 	debugInfoClient   debuginfo.Client
 	profilingDuration time.Duration
 	samplingRatio     float64
 }
 
-func NewManager(
+func NewProfilerPool(
 	logger log.Logger,
 	reg prometheus.Registerer,
+	ksymCache *ksym.Cache,
+	objCache objectfile.Cache,
 	writeClient profilestorepb.ProfileStoreServiceClient,
 	debugInfoClient debuginfo.Client,
 	profilingDuration time.Duration,
 	externalLabels model.LabelSet,
 	samplingRatio float64,
-) *Manager {
-	return &Manager{
+) *ProfilerPool {
+	return &ProfilerPool{
 		mtx:               &sync.RWMutex{},
-		profilerPools:     map[string]*ProfilerPool{},
+		activeTargets:     map[uint64]*Target{},
+		activeProfilers:   map[string]Profiler{},
+		externalLabels:    externalLabels,
 		logger:            logger,
 		reg:               reg,
-		externalLabels:    externalLabels,
-		ksymCache:         ksym.NewKsymCache(logger),
+		ksymCache:         ksymCache,
+		objCache:          objCache,
 		writeClient:       writeClient,
 		debugInfoClient:   debugInfoClient,
 		profilingDuration: profilingDuration,
 		samplingRatio:     samplingRatio,
 	}
 }
 
-func (m *Manager) Run(ctx context.Context, update <-chan map[string][]*Group) error {
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case targetSets := <-update:
-			err := m.reconcileTargets(ctx, targetSets)
-			if err != nil {
-				return err
-			}
-		}
-	}
-}
-
-func (m *Manager) reconcileTargets(ctx context.Context, targetSets map[string][]*Group) error {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	level.Debug(m.logger).Log("msg", "reconciling targets")
-	for name, targetSet := range targetSets {
-		pp, found := m.profilerPools[name]
-		if !found {
-			// An arbitrary coefficient. Number of assumed object files per target.
-			cacheSize := len(targetSet) * 5
-			pp = NewProfilerPool(
-				m.logger, m.reg,
-				m.ksymCache, objectfile.NewCache(cacheSize),
-				m.writeClient, m.debugInfoClient,
-				m.profilingDuration, m.externalLabels,
-				m.samplingRatio,
-			)
-			m.profilerPools[name] = pp
-		}
+func (pp *ProfilerPool) Profilers() map[string]Profiler {
+	pp.mtx.RLock()
+	defer pp.mtx.RUnlock()
 
-		pp.Sync(ctx, targetSet)
+	res := map[string]Profiler{}
+	for name, profiler := range pp.activeProfilers {
+		res[name] = profiler
 	}
-	return nil
+
+	return res
 }
 
-func (m *Manager) ActiveProfilers() map[string][]Profiler {
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
+func (pp *ProfilerPool) AddProfiler(ctx context.Context, profilerFunc NewProfilerFunc, allGroups func() map[int]model.LabelSet) error {
+	pp.mtx.Lock()
+	defer pp.mtx.Unlock()
 
-	profilerSet := map[string][]Profiler{}
-	for name, profilerPool := range m.profilerPools {
-		profilerSet[name] = profilerPool.Profilers()
-	}
+	labelSet := model.LabelSet{}
+	newProfiler := profilerFunc(
+		pp.logger,
+		pp.reg,
+		pp.ksymCache,
+		pp.objCache,
+		pp.writeClient,
+		pp.debugInfoClient,
+		labelSet,
+		pp.profilingDuration,
+		allGroups,
+	)
+	go func() {
```

Comment:
Maybe we should move this goroutine to the main and run it inside the rungroup, so we can stop everything else and clean up when a profiler fails. This is more align with our current code base.

Reply:
Good shout! Was planning to do this but forgot about it 😅 .

(I will address this and the rest of the comments after the review is complete 😄 )